text/0000-transparent-unions.md

```diff
@@ -0,0 +1,229 @@
+
+Like transparent `struct`s, transparent `union`s and `enum`s are FFI-safe if and only if their underlying representation type is also FFI-safe.
+
+A `union` may not be eligible for the same nonnull-style optimizations that a `struct` or `enum` (with the same fields) are eligible for. Adding `#[repr(transparent)]` to  `union` does not change this. To give a more concrete example, it is unspecified whether `size_of::<T>()` is equal to `size_of::<Option<T>>()`, where `T` is a `union` (regardless of whether it is transparent). The Rust compiler is free to perform this optimization if possible, but is not required to, and different compiler versions may differ in their application of these optimizations.  
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+`#[repr(transparent)]` on a `union` or `enum` is of limited use. There are cases where it is useful, but they're not common and some users might unnecessarily apply `#[repr(transparent)]` to a type in a cargo-cult fashion.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+It would be nice to make `MaybeUninit<T>` `#[repr(transparent)]`. This type is a `union`, and thus this RFC is required to allow making it transparent. One example in which a transparent representation would be useful is for unused parameters in an FFI-function:
+
+```rust
+#[repr(C)]
+struct Context {
+    // Imagine there a few fields here, defined by an external C library.
+}
+
+extern "C" fn log_event(message: core::ptr::NonNull<libc::c_char>,
+                        context: core::mem::MaybeUninit<Context>) {
+    // Log the message here, but ignore the context since we don't need it.
+}
+
+fn main() {
+    extern "C" {
+        fn set_log_handler(handler: extern "C" fn(core::ptr::NonNull<libc::c_char>,
+                                                  Context));
+    }
+
+    // Set the log handler so the external C library can call log_event.
+    unsafe {
+        // Transmuting is safe since MaybeUninit<Context> and Context
+        // have the same ABI.
+        set_log_handler(core::mem::transmute(log_event as *const ()));
+    }
+
+    // We can call it too. And since we don't care about the context and
+    // we're using MaybeUninit, we don't have to pay any extra cost for
+    // initializing something that's unused.
+    log_event(core::ptr::NonNull::new(b"Hello, world!\x00".as_ptr() as *mut _).unwrap(),
+              core::mem::MaybeUninit::uninitialized());
+}
+```
+
+It is also useful for consuming pointers to uninitialized memory:
+
+```rust
+#[repr(C)]
+struct Cryptor {
+    // Imagine there a few fields here, defined by an external C library.
+}
+
+// This function may be called from C (or Rust!), and matches the C
+// function signature: bool(Cryptor *cryptor)
+pub extern "C" fn init_cryptor(cryptor: &mut core::mem::MaybeUninit<Cryptor>) -> bool {
```

Comment:
> which the current implementation satisfies

That's an implementation detail that isn't explicitly guaranteed. Perhaps the guarantee was intended but never made explicitly. `repr(transparent)` will guarantee this, though.

Also, if `MaybeUninit` is ever made `repr(C)` (instead of `repr(transparent)`), then that guarantee no longer holds since C makes no guarantees here (though I don't think any sane implementation would add padding to a `union` with a single sized field).

Reply:
> That's an implementation detail that isn't explicitly guaranteed.

This was identified in the unsafe code guidelines repo as something that's probably worth guaranteeing. AFAICT nothing is blocking this - somebody would just need to write an RFC for this that works out the details. 

> Also, if MaybeUninit is ever made repr(C) 

Note that `MaybeUninit` contains a ZST, and C does not support ZSTs, so to make `MaybeUninit` `repr(C)` we would need to specify what does it mean for a ZST to be `repr(C)`. As you mention, there aren't many ways of doing that, which is a good thing. If there is only a single sane way of handling ZSTs in `repr(C)` or C FFI, then that's something that we could guarantee and IMO we should - we'd just need an RFC working out the details here.

---

My point is not that we don't need `repr(transparent)` for unions at all, but rather that there are some cross-cutting concerns between this RFC, and some other problems that we'd also like to solve (`repr(C)` ZSTs, union fields at offset 0 in some cases, etc.).